### PR TITLE
fix(processing): improve Thai and complex-script layout handling in gridProjection

### PR DESCRIPTION
**Pull request by unrealandychan**
## Summary

Thai text and other complex scripts (abugida/combining-diacritic scripts) were producing incorrect output from `gridProjection.ts` due to three interacting bugs. This PR fixes all three with targeted, backward-compatible changes.

---

### Bug 1 — `strLength` over-counted combining marks → skewed `medianWidth`

Thai vowels and tone marks (sara, mai, nikhahit …) are separate Unicode code-points stacked visually above or below a base consonant. The original code used:

```ts
strLength: bbox.str.length,
```

For the word `สวัสดี` (6 code-points, 2 of which are combining vowel signs), `str.length = 6` but only **4** are visible base characters. This inflated `strLength` caused `medianWidth = w / strLength` to be under-estimated, skewing column-gap detection, flowing-text indentation, and snap positions for every Thai page.

**Fix:** Replace `str.length` with `visibleCharCount(str)`, a new helper that skips Thai combining ranges (U+0E30–U+0E3A, U+0E47–U+0E4E) and general Unicode combining blocks.

---

### Bug 2 — Line-grouping split Thai tokens onto separate lines

Above-vowel signs (sara ii, mai ek/tho …) push the PDF bounding box upward, giving the item a lower `y` than neighbouring consonants on the same visual row. When such a token seeded a new line with a narrow `[lineMinY, lineMaxY]` band, the following base-consonant token had its midpoint fall just outside that band and was incorrectly placed on a new line.

**Fix:** Add a Thai-aware acceptance condition:
```ts
const thaiBottomInLine = containsThai(bbox.str) && bbox.y < lineMaxY + medianHeight * 0.5;
```
This allows a Thai item to join the current line when its top is within half a median line-height of the band bottom — catching the real-world stacked-diacritic case without affecting Latin content.

---

### Bug 3 — `isFlowingTextBlock` mis-classified Thai paragraphs as structured/tabular

Thai has no inter-word spaces. PDF engines emit each syllable cluster as a tiny separate bbox, so paragraph lines appear short relative to page width and fail the `FLOWING_WIDE_LINE_RATIO = 0.5` + `FLOWING_WIDE_LINE_THRESHOLD = 0.6` thresholds. The entire block was sent through the structured-grid path, which inserted large whitespace gaps between every cluster.

**Fix:** When any line in the block contains Thai text, apply relaxed thresholds:

| Threshold | Default | Thai |
|---|---|---|
| `wideLineRatio` | 0.5 | **0.3** |
| `wideLineThreshold` | 0.6 | **0.4** |

---

## New Exports

| Function | Description |
|---|---|
| `containsThai(str)` | Returns `true` if `str` contains any U+0E00–U+0E7F character |
| `visibleCharCount(str)` | Counts visible base characters, skipping combining diacritics |

Both are exported from `gridProjection.ts` and tested.

---

## Tests Added (`gridProjection.test.ts`)

**`containsThai`** — 5 cases: pure Thai, mixed Thai/Latin, pure Latin, empty, CJK

**`visibleCharCount`** — 5 cases: ASCII passthrough, Thai combining exclusion, combining-only string, Latin diacritics, plain Thai consonants

**`Thai bboxToLine`** — 3 cases:
- Multi-cluster Thai line with Y-shifted bboxes → correct single-line grouping
- Boundary case: combining-vowel token seeds narrow band, base consonant at exact `lineMaxY`
- Latin regression guard: genuinely different lines still split correctly

**Result: 139/139 tests pass, zero regressions.**

---

## Scope

Changes are confined to `src/processing/gridProjection.ts`. The fix is additive: all existing constants and logic paths are preserved; Thai-specific branches only activate when `containsThai()` returns true.

### CODE REVIEW

File: src/processing/gridProjection.ts [hunk 1]
```diff
@@ -4,6 +4,69 @@ import { cleanRawText } from "./cleanText.js";
...
+ * function is safe to use as a drop-in replacement everywhere.
+ */
+export function visibleCharCount(str: string): number {
+  let count = 0;
+  for (const ch of str) {
+    const cp = ch.codePointAt(0)!;
+    // Skip Thai tone marks & upper/lower vowel signs (combining diacritics)
+    // U+0E30–U+0E3A  lower vowels / short-a / sara
+    // U+0E40–U+0E44  leading vowels (written before the consonant)
+    // U+0E47–U+0E4E  maitaikhu, mai ek/tho/tri/jattawa, thanthacat, nikhahit
```
[the remark this answers is not in the document]
Comment (unrealandychan) [reply src/processing/gridProjection.ts:55]: @copilot apply changes based on [this feedback](https://github.com/run-llama/liteparse/pull/149#discussion_r3135685430)

File: src/processing/gridProjection.ts [hunk 2]
```diff
@@ -4,6 +4,69 @@ import { cleanRawText } from "./cleanText.js";
...
+    const isThaiCombining =
+      (cp >= 0x0e30 && cp <= 0x0e3a) || (cp >= 0x0e47 && cp <= 0x0e4e);
+    // General Unicode combining characters (Mn = non-spacing mark)
```
[the remark this answers is not in the document]
Comment (unrealandychan) [reply src/processing/gridProjection.ts:58]: @copilot apply changes based on [this feedback](https://github.com/run-llama/liteparse/pull/149#discussion_r3135685403)
